File: problem-3.java
// Time Complexity : O(n) where n is the length of the pattern string
// Space Complexity : O(n*m) where m is the length of each word in the s string
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap<Character, String> pMap = new HashMap<>();
        HashMap<String, Character> sMap = new HashMap<>();
        String[] str = s.split(" ");
        int plen = pattern.length();
        int slen = str.length;
        if(slen!=plen){
            return false;
        }
        for (int i=0; i<plen; i++){
            if (pMap.get(pattern.charAt(i)) == null){
                pMap.put(pattern.charAt(i), str[i]);
            }
            else{
                if(!pMap.get(pattern.charAt(i)).equals(str[i])){
                     return false;
                }
            }
            if (sMap.get(str[i]) == null){
                sMap.put(str[i], pattern.charAt(i));
            }
            else{
                if(sMap.get(str[i])!=pattern.charAt(i)){
                     return false;
                }
            }
            
        }
        return true;
    }
}

File: Problem3.cs
public class WordPattern
    {
        // Time Complexity : O(1) -- i think its constant time as the length of the entire string s is 3000 and t is 300
        // Space Complexity : O(1) -- since the max value in hashmap can be 26 key value pairs
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

                                        //abba  dog cat cat dog
        public bool WordPattern1(string pattern, string s)
        {
            if (s == null && pattern == null) return true;
            if (s == null || pattern == null) return false;

            string[] strArray = s.Split(' ');
            if (pattern.Length != strArray.Length) return false;

            Dictionary<char, string> patternMap = new Dictionary<char, string>();
            HashSet<string> strSet = new HashSet<string>();

            char[] patternArr = pattern.ToCharArray();

            for (int i = 0; i < patternArr.Length; i++)
            {
                //check if pattern character contains in hashmap
                if (patternMap.ContainsKey(patternArr[i]))
                {
                    if (patternMap[patternArr[i]] != strArray[i]) return false; //if pattern hashmap value and word in the string s does not match
                }
                else
                {
                    //check hashset contains the word in the string s
                    if (strSet.Contains(strArray[i]))
                    {
                        return false;
                    }
                    else
                    {
                        //add the pattern character and the word in the string s to dictionary
                        patternMap.Add(patternArr[i], strArray[i]);

                        //add the word in string s to the hashset
                        strSet.Add(strArray[i]);
                    }
                }
            }
            return true;
        }
    }
